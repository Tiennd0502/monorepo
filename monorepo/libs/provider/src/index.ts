export { default as TamaguiProvider } from './TamaguiProvider';
export { default as QueryClientProvider } from './QueryClientProvider';
