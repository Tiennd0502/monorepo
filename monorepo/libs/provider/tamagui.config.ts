import { createTamagui, createTokens } from 'tamagui';
import { config } from '@tamagui/config/v3';
import { themes } from '@tamagui/themes';

import {
  colorsMapping,
  radius,
  size,
  space,
  zIndex,
  defaultFont,
  headingFont,
} from '@monorepo/ui';

const tokens = {
  radius,
  size,
  space,
  zIndex,
};

const tamaguiConfig = createTamagui({
  ...config,
  defaultFont: 'body',
  themes,
  fonts: {
    body: defaultFont,
    heading: headingFont,
  },
  tokens: createTokens({
    ...tokens,
    color: colorsMapping,
  }),
});

type AppConfig = typeof config;

declare module 'tamagui' {
  type TamaguiCustomConfig = AppConfig;
}

export default tamaguiConfig;
