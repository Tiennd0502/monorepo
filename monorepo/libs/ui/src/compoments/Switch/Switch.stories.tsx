import { Meta, StoryObj } from '@storybook/react';

import Switch from '.';
import { Stack } from 'tamagui';

const meta: Meta<typeof Switch> = {
  title: 'Components/Switch',
  component: Switch,
  decorators: [
    (Story) => (
      <Stack width={450}>
        <Story />
      </Stack>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof Switch>;

export const Default: Story = {
  args: {
    children: 'Default',
  },
};

export const Checked: Story = {
  args: {
    defaultChecked: true,
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
  },
};
