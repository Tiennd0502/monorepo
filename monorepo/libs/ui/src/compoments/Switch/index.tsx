import {
  Label,
  Stack,
  XStack,
  YStack,
  styled,
  createSwitch,
  SwitchProps as TSwitchProps,
} from 'tamagui';
import { useId } from 'react';

const Frame = styled(Stack, {
  width: '$10',
  height: '$6',
  borderRadius: '$3.75',
  borderWidth: 0,
  padding: '$0.5',

  variants: {
    checked: {
      indeterminate: {},
      true: {
        backgroundColor: '$successPrimary',
      },
      false: {
        backgroundColor: '$backgroundDisabled',
      },
    },

    disabled: {
      true: {
        opacity: 0.6,
      },
    },
  } as const,

  defaultVariants: {
    checked: false,
  },
});

const Thumb = styled(Stack, {
  width: '$5',
  height: '$5',
  backgroundColor: '$secondary',
  borderRadius: '$5',

  variants: {
    checked: {
      true: {},
      false: {
        opacity: 0.8,
      },
    },
  } as const,
});

export const SwitchBase = createSwitch({
  Frame,
  Thumb,
});

interface SwitchProps extends Omit<TSwitchProps, ''> {
  label?: string;
}

const Switch = ({ label, ...props }: SwitchProps) => {
  const id = useId();

  return (
    <YStack width={200} alignItems="center" gap="$3">
      <XStack gap="$2" alignItems="center">
        {label && (
          <Label htmlFor={id} color="$textDefault">
            {label}
          </Label>
        )}
        <SwitchBase id={id} {...props}>
          <SwitchBase.Thumb animation="quick" />
        </SwitchBase>
      </XStack>
    </YStack>
  );
};

export default Switch;
