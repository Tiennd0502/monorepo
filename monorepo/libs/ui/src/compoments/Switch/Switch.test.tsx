import {fireEvent, render, waitFor} from '@testing-library/react-native';

import { TamaguiProvider } from '@monorepo/provider';
import Switch from '.';

const onCheckedChange = jest.fn();

const baseProps = {
  label: 'Label',
  onCheckedChange
};

const setup = (moreProps = {}) => {
  const props = {
    ...baseProps,
    ...moreProps,
  };
  
  return render(
    <TamaguiProvider>
      <Switch {...props} />
    </TamaguiProvider>
  );
};

describe('Switch', () => {
  window.setImmediate = window.setTimeout

  it('should render Switch component', () => {
    const container = setup();

    expect(container).toMatchSnapshot();
  });

  it('should render Switch with checked true', () => {
    const container = setup({defaultChecked: true});

    expect(container).toMatchSnapshot();
  });

  it('calls onCheckedChange when being clicked', () => {
    const onCheckedChange = jest.fn();
    const {getByTestId} = setup({onCheckedChange});

    waitFor(() => {
      const label = getByTestId('switch');
      fireEvent.press(label);
  
      expect(onCheckedChange).toHaveBeenCalled();
    })
  });
});
