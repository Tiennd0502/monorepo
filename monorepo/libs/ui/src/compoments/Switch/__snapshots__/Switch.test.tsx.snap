// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render Switch component 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "gap": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 8,
      }
    }
  >
    <Text
      htmlFor=":r0:"
      id=":r1:"
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#999",
          "display": "flex",
          "fontFamily": "NunitoSans_7pt-Medium",
          "fontSize": 16,
          "letterSpacing": -0.5,
          "lineHeight": 50,
        }
      }
      suppressHighlighting={true}
      userSelect="none"
    >
      Label
    </Text>
    <View
      accessibilityRole="switch"
      accessibilityState={
        {
          "checked": false,
        }
      }
      id=":r0:"
      size="$true"
      style={
        {
          "backgroundColor": "#909090",
          "borderBottomLeftRadius": 15,
          "borderBottomRightRadius": 15,
          "borderBottomWidth": 0,
          "borderLeftWidth": 0,
          "borderRightWidth": 0,
          "borderStyle": "solid",
          "borderTopLeftRadius": 15,
          "borderTopRightRadius": 15,
          "borderTopWidth": 0,
          "height": 24,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 40,
        }
      }
      testID="switch"
      unstyled={false}
    >
      <View
        onLayout={[Function]}
        style={
          {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          onLayout={[Function]}
          size="$true"
          style={
            {
              "alignSelf": "flex-start",
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderBottomLeftRadius": 20,
              "borderBottomRightRadius": 20,
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "height": 20,
              "transform": [
                {
                  "translateX": 0,
                },
              ],
              "width": 20,
            }
          }
          unstyled={false}
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Switch should render Switch with checked true 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "gap": 12,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 8,
      }
    }
  >
    <Text
      htmlFor=":r2:"
      id=":r3:"
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#999",
          "display": "flex",
          "fontFamily": "NunitoSans_7pt-Medium",
          "fontSize": 16,
          "letterSpacing": -0.5,
          "lineHeight": 50,
        }
      }
      suppressHighlighting={true}
      userSelect="none"
    >
      Label
    </Text>
    <View
      accessibilityRole="switch"
      accessibilityState={
        {
          "checked": true,
        }
      }
      id=":r2:"
      size="$true"
      style={
        {
          "backgroundColor": "#27ae60",
          "borderBottomLeftRadius": 15,
          "borderBottomRightRadius": 15,
          "borderBottomWidth": 0,
          "borderLeftWidth": 0,
          "borderRightWidth": 0,
          "borderStyle": "solid",
          "borderTopLeftRadius": 15,
          "borderTopRightRadius": 15,
          "borderTopWidth": 0,
          "height": 24,
          "paddingBottom": 2,
          "paddingLeft": 2,
          "paddingRight": 2,
          "paddingTop": 2,
          "width": 40,
        }
      }
      testID="switch"
      unstyled={false}
    >
      <View
        onLayout={[Function]}
        style={
          {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          onLayout={[Function]}
          size="$true"
          style={
            {
              "alignSelf": "flex-end",
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderBottomLeftRadius": 20,
              "borderBottomRightRadius": 20,
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "height": 20,
              "transform": [
                {
                  "translateX": 0,
                },
              ],
              "width": 20,
            }
          }
          unstyled={false}
        />
      </View>
    </View>
  </View>
</View>
`;
